crates/handlers/src/upstream_oauth2/logout.rs

```diff
@@ -0,0 +1,148 @@
+
+    #[error("session was not found")]
+    SessionNotFound,
+}
+
+impl_from_error_for_route!(mas_storage::RepositoryError);
+
+impl From<reqwest::Error> for RouteError {
+    fn from(err: reqwest::Error) -> Self {
+        Self::Internal(Box::new(err))
+    }
+}
+
+/// Get RP-initiated logout URLs for a user's upstream providers
+///
+/// This retrieves logout endpoints from all connected upstream providers that
+/// support RP-initiated logout.
+///
+/// # Parameters
+///
+/// * `repo`: The repository to use
+/// * `url_builder`: URL builder for constructing redirect URIs
+/// * `cookie_jar`: Cookie from user's browser session
+///
+/// # Returns
+///
+/// Information about upstream logout endpoints the user should be redirected to
+///
+/// # Errors
+///
+/// Returns a `RouteError` if there's an issue accessing the repository
+pub async fn get_rp_initiated_logout_endpoints<E>(
+    url_builder: &UrlBuilder,
+    repo: &mut impl RepositoryAccess<Error = E>,
+    cookie_jar: &CookieJar,
+) -> Result<UpstreamLogoutInfo, RouteError>
+where
+    RouteError: std::convert::From<E>,
+{
+    let mut result: UpstreamLogoutInfo = UpstreamLogoutInfo::default();
+    // Set the post-logout redirect URI to our app's logout completion page
+    let post_logout_redirect_uri = url_builder
+        .absolute_url_for(&mas_router::Login::default())
+        .to_string();
+    result.post_logout_redirect_uri = Some(post_logout_redirect_uri.clone());
+
+    let sessions_cookie = UpstreamSessionsCookie::load(cookie_jar);
+    // Standard location for OIDC end session endpoint
+    let session_ids = sessions_cookie.session_ids();
+    if session_ids.is_empty() {
+        return Ok(result);
+    }
+    // We only support the first upstream session
+    let mut provider = None;
+    let mut upstream_session = None;
+    for session_id in session_ids {
+        // Get the session and assign its value, wrapped in Some
+        let session = repo
+            .upstream_oauth_session()
+            .lookup(session_id)
+            .await?
+            .ok_or(RouteError::SessionNotFound)?;
+        // Get the provider and assign its value, wrapped in Some
+        let prov = repo
+            .upstream_oauth_provider()
+            .lookup(session.provider_id)
+            .await?
+            .ok_or(RouteError::ProviderNotFound)?;
+
+        if prov.allow_rp_initiated_logout {
+            upstream_session = Some(session);
+            provider = Some(prov);
+            break;
+        }
+    }
+
+    // Check if we found a provider with allow_rp_initiated_logout
+    if let Some(provider) = provider {
+        // Look for end session endpoint
+        // In a real implementation, we'd have end_session_endpoint fields in the
+        // provider For now, we'll try to construct one from the issuer if
+        // available
+        if let Some(issuer) = &provider.issuer {
+            let end_session_endpoint = format!("{issuer}/protocol/openid-connect/logout");
```

Comment:
So that I don't forget: this must use the `end_session_endpoint` metadata. Please also make sure to allow overriding this, like all the other endpoints